File: LC/[2597] The-Number-of-Beautiful-Subsets-DFS.py
class Solution:
    def beautifulSubsets(self, nums: List[int], k: int) -> int:
        N = len(nums)

        def dfs(st):

            if st >= N:
                return 1

            has_prior_diff_k = False
            for ii in range(st):
                if ii in status and (abs(nums[ii]-nums[st]) == k):
                    has_prior_diff_k = True
                    break

            if has_prior_diff_k:
                return dfs(st+1)
            else:
                no_take = dfs(st+1)
                status.add(st)
                take = dfs(st+1)
                status.remove(st)
                return take+no_take

        status = set()
        return dfs(0)-1

File: LC/[2597] The-Number-of-Beautiful-Subsets-DFS.cpp
class Solution {
public:
    int beautifulSubsets(vector<int>& nums, int k) 
    {
        return dfs(0, 0, nums, k) - 1;
    }

    int dfs(int cur, int state, vector<int>& nums, int k)
    {
        if (cur==nums.size()) return 1;

        int has_prior_diff_k = 0;
        for (int i=0; i<cur; i++)
        {
            if ((state>>i)&1 && abs(nums[i]-nums[cur])==k)
            {
                has_prior_diff_k = 1;
                break;
            }
        }

        if (has_prior_diff_k == 1){
            return dfs(cur+1, state, nums, k);
        }else{
            return dfs(cur+1, state, nums, k) + dfs(cur+1, state+(1<<cur), nums, k);
        }
    }
};